### DISCUSSION TITLE
Kernel modules

### DISCUSSION THREAD

**Discussion by consp** [xblax/flashforge_ad5m_klipper_mod#66] [General]
@xblax you mentiond kernel module compilation somewhere, but I could not find the post you mentioned it (or I was imagining it). Did you have any luck compiling a module? 

I wanted to check if I maybe could compile uinput (which would make it possible to use tslib instead of evdev for touchscreen input).

Unfortunately I'm getting segfaults all over so I've missed something somewhere and it's been a while with compiling for existing kernels. Probably since I used mainline 5.4.61 and not the special sunxi version for the t113 but I have no clue where to get my hands on that one.

Some help would be appreciated :)

**Comment by xblax** [comment]
Yes, I compiled some of the crypto modules needed for WiFi. They are in the buildroot overlay.

This should be the correct Kernel sources: https://github.com/Tina-Linux/linux-5.4
Was originally posted [here](https://github.com/g992/flashforge-ad5m-5mpro-research/issues/6#issuecomment-1934927466) by @jlodew0 

And I used that together with the linaro 7.4.1 toolchain which was used to compile the stock buildroot: 
https://releases.linaro.org/components/toolchain/binaries/7.4-2019.02/arm-linux-gnueabi/

Probably our buildroot toolchain would work as well, have not tested it.

Make sure to NOT build the Kernel from the git checkout, it needs to be built from a source tree without .git 
Otherwise you will get his issue:
`sha512_generic: version magic '5.4.61+ SMP preempt mod_unload ARMv7 p2v8 ' should be '5.4.61 SMP preempt mod_unload ARMv7 p2v8 '`

Original kernel config is available in `/proc/config.gz`

For cross-compiling you just need to supply the CROSS_COMPILE parameter to `make` like this:
```
make ARCH=arm CROSS_COMPILE=~/build/gcc-linaro-7.4.1-2019.02-x86_64_arm-linux-gnueabi/bin/arm-linux-gnueabi- menuconfig
make ARCH=arm CROSS_COMPILE=~/build/gcc-linaro-7.4.1-2019.02-x86_64_arm-linux-gnueabi/bin/arm-linux-gnueabi- modules
```

**Comment by consp** [comment]
I knew I read it somewhere, just forgot where!

Tnx for the help! The gist of it is that I simply had the wrong sources (mainline), everything else was the same. Should have known mainline wasn't going to work but it was late in the evening and I was not thinking properly.

At least it's now also documented here!

> Probably our buildroot toolchain would work as well, have not tested it.

I suspect problems, eabi vs eabihf is likely an issue unless that has changed in the past few years.

> And I used that together with the linaro 7.4.1 toolchain which was used to compile the stock buildroot:

I'll see what the difference is with the 5.x version the config tells me it's build with. Don't expect too much differences though but it might be worth a check.

> Make sure to NOT build the Kernel from the git checkout, it needs to be built from a source tree without .git

CONFIG_LOCALVERSION_AUTO=n and empty localversion should be enough I think. Removing .git also works of course.

**Reply by consp** [reply]
Works now, was simply getting the correct source...

I can confirm the linaro 5.3.1-2016-05 compilers also work. uinput has some quirks as it segv's sunxi-ts if you load/unload it too often but that's not uinput's fault.
